ergo/distributions/point_density.py

```diff
@@ -0,0 +1,241 @@
+
+    def ppf(self, q):
+        low_idx = np.argmax(self.cumulative_normed_ps >= q)
+        high_idx = low_idx + 1
+        low_x = self.normed_xs[low_idx]
+        high_x = self.normed_xs[np.minimum(high_idx, self.normed_xs.size - 1)]
+        low_cum = np.where(low_idx==0, 0, self.cumulative_normed_ps[low_idx-1])
+        high_cum = self.cumulative_normed_ps[high_idx-1]
+        dist = high_cum - low_cum
+        # print(f'lx: {low_x} lc: {low_cum} hx: {high_x} hc: {high_cum}')
+        normed_x = np.where(dist == 0, low_x, (q - low_cum) / dist * high_x + (high_cum - q) / dist * low_x)
+        return self.scale.denormalize_point(normed_x)
+
+    def sample(self):
+        raise NotImplementedError
+
+    # Scaled
+
+    def normalize(self):
+        return PointDensity(
+            self.normed_xs, self.normed_densities, scale=Scale(0, 1), normalized=True
+        )
+
+    def denormalize(self, scale: Scale):
+        return PointDensity(
+            self.normed_xs, self.normed_densities, scale=scale, normalized=True
+        )
+
+    # Structured
+
+    @classmethod
+    def structure(cls, params):
+        class_params, numeric_params = params
+        density_class, scale_classes = class_params
+        density_numeric, scale_numeric = numeric_params
+        return density_class(
+            xs=density_numeric[0],
+            densities=density_numeric[1],
+            scale=scale_classes[0].structure((scale_classes, scale_numeric)),
+            normalized=True,
+            traceable=True,
+            cumulative_normed_ps=density_numeric[2],
+        )
+
+    def destructure(self):
+        scale_classes, scale_numeric = self.scale.destructure()
+        class_params = (self.__class__, scale_classes)
+        self_numeric = self.normed_xs, self.normed_densities, self.cumulative_normed_ps
+        numeric_params = (self_numeric, scale_numeric)
+        return (class_params, numeric_params)
+
+    # Optimizable
+
+    @classmethod
+    def from_conditions(cls, *args, fixed_params=None, scale=None, **kwargs):
+        if scale is None:
+            # TODO: Should we do this?
+            scale = Scale(0, 1)
+        if fixed_params is None:
+            # TODO: Seems weird to denormalize when will be normalized in normalize_fixed_params
+            fixed_params = {
+                "xs": scale.denormalize_points(np.linspace(
+                    0, 1, constants.point_density_default_num_points
+                ))
+            }
+        return super(PointDensity, cls).from_conditions(
+            *args, fixed_params=fixed_params, scale=scale, **kwargs
+        )
+
+    @classmethod
+    def from_params(cls, fixed_params, opt_params, scale=None, traceable=True):
+        # TODO: traceable is always True here
+        if scale is None:
+            # TODO: Should we do this?
+            scale = Scale(0, 1)
+        xs = fixed_params["xs"]
+        # ps = np.abs(opt_params)
+        ps = nn.softmax(opt_params)
```

Comment:
Not sure if abs vs. softmax really matters here. I was experimenting with changing abs to softmax but it didn't make much of a difference as far as I could tell.